enhancements/machine-api/cluster-autoscaler-integration.md

```diff
@@ -0,0 +1,449 @@
+see-also:
+  - "/enhancements/machine-api/cluster-autoscaler-operator.md"  
+replaces:
+superseded-by:
+---
+
+# Cluster Autoscaler Integration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is a tool that users
+can deploy to a cluster to enable the automatic provisioning of extra resources when workloads exceed the cluster's
+capacity.
+It can also be configured to scale down resources if they are surplus to current requirement.
+
+In normal operation, a user would configure the Cluster Autoscaler to talk directly to their cloud provider and create
+new instances by expanding a nodegroup, such as an Autoscaling Group on AWS EC2.
+Because OpenShift users the Machine-API to create compute resources, OpenShift clusters do not have nodegroups in the
+way that the Cluster Autoscaler expects.
+
+This proposal details how a new provider will be implemented within the Cluster Autoscaler to integrate with Machine API
+and the upstream [Cluster API](https://github.com/kubernetes-sigs/cluster-api) project, by scaling MachineSets and
+MachineDeployments to create new capacity within an OpenShift or Kubernetes Cluster API managed cluster.
+
+## Motivation
+
+### Goals
+
+- Enable automatic expansion of OpenShift cluster resources when capacity is exceeded
+- Enable automatic contraction of OpenShift cluster resources when capacity is under-utilised
+- Provide an agnostic autoscaler integration that will work with different implementations and versions of Cluster API
+
+### Non-Goals
+
+- Install the Cluster Autoscaler on OpenShift clusters, this will be handled by the Cluster Autoscaler Operator
+- Add support for MachineDeployments to OpenShift, though the Cluster Autoscaler will support them
+
+## Proposal
+
+Create a new Cloud Provider integration within the cluster-autoscaler called `clusterapi`.
+
+This integration will be implemented in the form of a new Cloud Provider within the Cluster Autoscaler.
+The Cloud Provider will use MachineSets and MachineDeployments that have opted in to autoscaling to provide the
+Cluster Autoscaler with information about the resources within a cluster and the resources available for scaling.
+
+MachineSets and MachineDeployments will be opted-in to autoscaling if they are annotated with two annotations, which
+denote the maximum and minimum sizes that the autoscaler should scale the Node group within, eg:
+```yaml
+metadata:
+  annotations:
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-max-size": "6"
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-min-size": "1"
+```
```

Comment:
The telco/edge hardware team has something similar in the cluster-api-provider-baremetal for automatically provisioning hosts when new inventory data is provided. It would be good to coordinate so that we don't end up with 2 controllers competing to set the replica count of a MachineSet.

https://github.com/openshift/cluster-api-provider-baremetal/pull/52

/cc @mhrivnak

Reply:
I would imagine users wouldn't be enabling the cluster autoscaler on baremetal, do you think we should be actively preventing this somehow?

Reply:
It seems unlikely that the autoscaler would be as much use on bare metal, although theoretically it would work if you had unused inventory available. I don't know if it would be a good policy to disable it completely for the platform. Do we disable other services based on the platform? Every time some unique aspect of bare metal comes up, the goal is to make it work the same as everywhere else.

There are ways we could make the two things compatible. We could make the bare metal provider honor the scaling limits for a MachineSet, for example. Or we could use an annotation to designate one or the other scaling tool as being responsible for scaling a given MachineSet so that they never compete. Both of those probably need more thought about how to implement them and whether they actually solve the problem.

Reply:
It's hard to imagine a bare metal cluster using load-based autoscaling of MachineSets. I guess it would have two potential goals: use less electricity, or actually shuffle hosts among clusters based on which cluster needs capacity at the time. Just the latency involved in provisioning new hosts diminishes the value, and we'd need additional management capabilities to support moving hosts among clusters.

Without a significant investment in providing a good experience for load-based autoscaling on bare metal, it seems best to at least discourage it if not outright prevent it.

Prevention could be done in at least these two ways:
1. Have the autoscaler look at a MacineSet's ProviderSpec and decline to scale it if it finds a BareMetalProviderSpec.
2. Add a validating admission webhook in CAPBM that prevents those annotations from being added to a MachineSet.

Or if we want to allow either but not both kinds of autoscaling, an admission webhook could enforce that only one or the other type of annotation could exist on a MachineSet.

Reply:
i think this is good discussion, but i'd like to gently point out that this merge request is about documenting the current behavior of the cluster autoscaler operator and its integration.

i would love to see a followup where we capture the discussion for future improvements to the CAO.